library/bignum.c

```diff
@@ -1249,27 +1249,34 @@ int mbedtls_mpi_sub_abs( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
     return( ret );
 }
 
-/*
- * Signed addition: X = A + B
+/* Common function for signed addition and subtraction.
+ * Calculate A + B * flip_B where flip_B is 1 or -1.
  */
-int mbedtls_mpi_add_mpi( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi *B )
+static int add_sub_mpi( mbedtls_mpi *X,
+                        const mbedtls_mpi *A, const mbedtls_mpi *B,
+                        int flip_B )
 {
     int ret, s;
     MPI_VALIDATE_RET( X != NULL );
     MPI_VALIDATE_RET( A != NULL );
     MPI_VALIDATE_RET( B != NULL );
 
     s = A->s;
-    if( A->s * B->s < 0 )
+    if( A->s * B->s * flip_B < 0 )
```

Comment:
Yes, although I'm always wary of bitwise operations with negative numbers since they can create trap representations, at least in theory. And I'm not convinced this is worth improving, especially in 2.28. And if it is, it's preexisting and feels like scope creep here.

Reply:
If we change it, I think we should change it everywhere there's a sign multiplication. The current code might be slightly suboptimal in performance, but it's at least consistent.